File: web/src/blunderTagger.ts
// Chegga Web — WHY a mistake/blunder happened (Phase 2)
//
// Ported from Chegga's own `app/services/blunder_tagger.py::tag_move`, same
// priority order and same "hanging" approximation (undefended, or defended
// only by pieces worth more than a cheaper attacker) — not a full
// static-exchange evaluator, deliberately, per the backend's own note.

import { Chess, type Color, type PieceSymbol, type Square } from "chess.js";

export const MISSED_MATE = "missed_mate"; // a forced mate was available and you didn't play it
export const ALLOWED_MATE = "allowed_mate"; // your move let the opponent force mate
export const HUNG_MATERIAL = "hung_material"; // a piece of yours is now capturable for less than it's worth
export const MISSED_CAPTURE = "missed_capture"; // the engine's best move was a capture and you played something else
export const POSITIONAL = "positional"; // none of the above

const PIECE_VALUES: Record<PieceSymbol, number> = {
  p: 1,
  n: 3,
  b: 3,
  r: 5,
  q: 9,
  k: 100, // never actually hangs, but keeps the value lookup total
};

const TAGGABLE_CLASSIFICATIONS = new Set(["mistake", "blunder"]);

function isHanging(board: Chess, square: Square, pieceColor: Color): boolean {
  const piece = board.get(square);
  if (!piece || piece.type === "k") return false;

  const opponent: Color = pieceColor === "w" ? "b" : "w";
  const attackers = board.attackers(square, opponent);
  if (attackers.length === 0) return false;

  const defenders = board.attackers(square, pieceColor);
  if (defenders.length === 0) return true;

  const cheapestAttacker = Math.min(...attackers.map((sq) => PIECE_VALUES[board.get(sq)!.type]));
  return cheapestAttacker < PIECE_VALUES[piece.type];
}

/** Exported for reuse beyond blunder-tagging: the vision-drill trainer
 * ("is anything hanging?") and the optional live hang-warning during bot
 * games both want the same "is this side about to lose material for
 * free" check, not a reimplementation of it. */
export function hasHangingPiece(board: Chess, color: Color): boolean {
  for (const row of board.board()) {
    for (const cell of row) {
      if (cell && cell.color === color && isHanging(board, cell.square, color)) return true;
    }
  }
  return false;
}

export interface TagMoveArgs {
  fenBefore: string;
  uci: string;
  san: string;
  bestMoveSan?: string;
  evalBeforeMate?: number;
  evalAfterMate?: number;
  sideToMove: "white" | "black";
  classification: string;
}

export function tagMove(args: TagMoveArgs): string | undefined {
  if (!TAGGABLE_CLASSIFICATIONS.has(args.classification)) return undefined;

  const moverSign = args.sideToMove === "white" ? 1 : -1;
  const bestBeforeMoverMate = args.evalBeforeMate !== undefined ? args.evalBeforeMate * moverSign : undefined;
  const afterMoverMate = args.evalAfterMate !== undefined ? args.evalAfterMate * moverSign : undefined;

  if (bestBeforeMoverMate !== undefined && bestBeforeMoverMate > 0) return MISSED_MATE;
  if (afterMoverMate !== undefined && afterMoverMate < 0) return ALLOWED_MATE;

  const board = new Chess(args.fenBefore);
  const moverColor = board.turn();
  board.move({ from: args.uci.slice(0, 2), to: args.uci.slice(2, 4), promotion: args.uci.slice(4) || undefined });
  if (hasHangingPiece(board, moverColor)) return HUNG_MATERIAL;

  if (args.bestMoveSan?.includes("x") && !args.san.includes("x")) return MISSED_CAPTURE;

  return POSITIONAL;
}

File: archive/backend/app/services/blunder_tagger.py
"""Classifies WHY a mistake/blunder happened, not just how costly it was.
centipawn_loss + classification already say "this move lost ~150cp";
blunder_tag says "...because you hung your knight" -- the difference
between a stat and something a drill or a coaching report can act on.

Tags are mutually exclusive, checked in priority order (most diagnostic
first): a hung piece is a more useful lesson than "positional" even if a
position also happens to be slightly worse for other reasons.

This is deliberately NOT a full static-exchange evaluator -- "hanging"
below is an approximation (undefended, or defended only by pieces worth
more than a cheaper attacker) that catches the common, teachable cases
without another engine call per move.
"""
import chess

MISSED_MATE = "missed_mate"       # a forced mate was available and you didn't play it
ALLOWED_MATE = "allowed_mate"     # your move let the opponent force mate
HUNG_MATERIAL = "hung_material"   # a piece of yours is now capturable for less than it's worth
MISSED_CAPTURE = "missed_capture"  # the engine's best move was a capture and you played something else
POSITIONAL = "positional"          # none of the above -- lost eval without a clean material/tactical story

_PIECE_VALUES = {
    chess.PAWN: 1,
    chess.KNIGHT: 3,
    chess.BISHOP: 3,
    chess.ROOK: 5,
    chess.QUEEN: 9,
    chess.KING: 100,  # never actually hangs, but keeps the value lookup total
}

_TAGGABLE_CLASSIFICATIONS = {"mistake", "blunder"}


def _is_hanging(board: chess.Board, square: chess.Square, piece_color: bool) -> bool:
    piece = board.piece_at(square)
    if piece is None or piece.piece_type == chess.KING:
        return False
    attackers = board.attackers(not piece_color, square)
    if not attackers:
        return False
    defenders = board.attackers(piece_color, square)
    if not defenders:
        return True
    cheapest_attacker = min(_PIECE_VALUES[board.piece_at(sq).piece_type] for sq in attackers)
    return cheapest_attacker < _PIECE_VALUES[piece.piece_type]


def _has_hanging_piece(board: chess.Board, color: bool) -> bool:
    return any(
        _is_hanging(board, square, color)
        for square, piece in board.piece_map().items()
        if piece.color == color
    )


def tag_move(
    fen_before: str,
    uci: str,
    san: str,
    best_move_san: str | None,
    eval_before_mate: int | None,
    eval_after_mate: int | None,
    side_to_move: str,
    classification: str,
) -> str | None:
    if classification not in _TAGGABLE_CLASSIFICATIONS:
        return None

    mover_sign = 1 if side_to_move == "white" else -1
    best_before_mover_mate = eval_before_mate * mover_sign if eval_before_mate is not None else None
    after_mover_mate = eval_after_mate * mover_sign if eval_after_mate is not None else None

    if best_before_mover_mate is not None and best_before_mover_mate > 0:
        return MISSED_MATE
    if after_mover_mate is not None and after_mover_mate < 0:
        return ALLOWED_MATE

    board = chess.Board(fen_before)
    mover_color = board.turn
    board.push(chess.Move.from_uci(uci))
    if _has_hanging_piece(board, mover_color):
        return HUNG_MATERIAL

    if best_move_san and "x" in best_move_san and "x" not in san:
        return MISSED_CAPTURE

    return POSITIONAL
